docs/examples/bayesian_optimisation.py

```diff
@@ -0,0 +1,737 @@
+
+# %% [markdown]
+# First we'll specify the domain over which we wish to optimise the function, as well as
+# sampling some initial points for fitting our surrogate model using a space-filling design.
+
+# %%
+lower_bound = jnp.array([0.0])
+upper_bound = jnp.array([1.0])
+initial_sample_num = 5
+
+initial_x = tfp.mcmc.sample_halton_sequence(
+    dim=1, num_results=initial_sample_num, seed=key, dtype=jnp.float64
+).reshape(-1, 1)
+initial_y = forrester(initial_x)
+D = gpx.Dataset(X=initial_x, y=initial_y)
+
+
+# %% [markdown]
+# Next we'll define our GP model in the usual way, using a Matérn52 kernel, and fit the
+# kernel parameters by minimising the negative log-marginal likelihood. We'll wrap this in
+# a function as we'll be repeating this process at each iteration of the BO loop.
+
+
+# %%
+def generate_optimised_posterior(
+    data: gpx.Dataset, prior: gpx.Module, key: Array
+) -> gpx.Module:
+    likelihood = gpx.Gaussian(
+        num_datapoints=data.n, obs_noise=jnp.array(1e-6)
+    )  # Our function is noise-free, so we set the observation noise to a very small value
+    likelihood = likelihood.replace_trainable(obs_noise=False)
+
+    posterior = prior * likelihood
+
+    negative_mll = gpx.objectives.ConjugateMLL(negative=True)
+    negative_mll(posterior, train_data=data)
+    negative_mll = jit(negative_mll)
+
+    opt_posterior, history = gpx.fit(
+        model=posterior,
+        objective=negative_mll,
+        train_data=data,
+        optim=ox.adam(learning_rate=0.01),
+        num_iters=1000,
+        safe=True,
+        key=key,
+        verbose=False,
+    )
+
+    return opt_posterior
+
+
+mean = gpx.mean_functions.Zero()
+kernel = gpx.kernels.Matern52()
+prior = gpx.Prior(mean_function=mean, kernel=kernel)
+opt_posterior = generate_optimised_posterior(D, prior, key)
+
+# %% [markdown]
+# We can then sample a function from the posterior distribution of the surrogate model. We
+# will do this using the `sample_approx` method, which generates an approximate sample
+# from the posterior using decoupled sampling introduced in ([Wilson et al.,
+# 2020](https://proceedings.mlr.press/v119/wilson20a.html)) and discussed in our [Pathwise
+# Sampling Notebook](https://docs.jaxgaussianprocesses.com/examples/spatial/). This method
+# is used as it enables us to sample from the posterior in a manner which scales linearly
+# with the number of points sampled, $O(N)$, mitigating the cubic cost associated with
+# drawing exact samples from a GP posterior, $O(N^3)$. It also generates more accurate
+# samples than many other methods for drawing approximate samples from a GP posterior.
+#
+# Note that we also define a `utility_fn` which calls the approximate
+# sample but returns the value returned as a scalar. This is because the `sample_approx`
+# function returns an array of shape $[N, B]$, with $N$ being the number of points within
+# each sample and $B$ being the number of samples drawn. We'll only be drawing (and
```

Comment:
this is definitely something we will hide away once we have some BO functionality